backend/src/Database.hs

```diff
@@ -76,41 +85,69 @@ insertPlayerIfNotExists name = do
   liftIO . withResource env.dbPool $ \conn -> do
     case player of
       Nothing -> do
-        execute conn "INSERT INTO Players (name, country) VALUES (?, ?)" (WritePlayer {name, country = Nothing})
-        rowId <- query_ conn "SELECT last_insert_rowid();"
-        case rowId of
-          [Only pid] -> pure pid
-          _ -> pure 0
-      Just p -> pure p.pid
+        let insertPlayerQuery = "INSERT INTO Players (name, country) VALUES (?, ?) RETURNING id"
+        readSingle $ query conn insertPlayerQuery (WritePlayer {name, country = Nothing})
+      Just (ReadPlayer {pid}) -> pure pid
 
-insertGameReport :: WriteProcessedGameReport -> AppM ()
+insertGameReport :: WriteProcessedGameReport -> AppM ReadProcessedGameReport
 insertGameReport report = do
   env <- ask
   liftIO . withResource env.dbPool $ \conn -> do
-    execute
-      conn
-      "INSERT INTO GameReports (\
-      \timestamp,\
-      \winner,\
-      \loser,\
-      \side,\
-      \victory,\
-      \match,\
-      \competition,\
-      \league,\
-      \expansions,\
-      \treebeard,\
-      \actionTokens,\
-      \dwarvenRings,\
-      \turns,\
-      \corruption,\
-      \mordor,\
-      \initialEyes,\
-      \aragornTurn,\
-      \strongholds,\
-      \interestRating,\
-      \comments,\
-      \winnerRatingAfter,\
-      \loserRatingAfter\
-      \) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"
-      report
\ No newline at end of file
+    let insertReportQuery =
+          "INSERT INTO GameReports (\
+          \timestamp,\
+          \winner,\
+          \loser,\
+          \side,\
+          \victory,\
+          \match,\
+          \competition,\
+          \league,\
+          \expansions,\
+          \treebeard,\
+          \actionTokens,\
+          \dwarvenRings,\
+          \turns,\
+          \corruption,\
+          \mordor,\
+          \initialEyes,\
+          \aragornTurn,\
+          \strongholds,\
+          \interestRating,\
+          \comments,\
+          \winnerRatingAfter,\
+          \loserRatingAfter\
+          \) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)\
+          \RETURNING id"
+    rid <- readSingle $ query conn insertReportQuery report
+
+    let nameQuery = "SELECT name FROM Players WHERE id = ?"
+    winner <- readSingle $ query conn nameQuery (Only report.winner)
+    loser <- readSingle $ query conn nameQuery (Only report.loser)
+
+    pure
+      ReadProcessedGameReport
+        { rid,
+          timestamp = report.timestamp,
+          winner,
```

Comment:
Have you seen this pattern before? You're allowed to write this, instead of something like `winner = winner` as long as the field of the record has the same name as the variable in question. It's called a **record pun**. Neat convenience feature.

https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/record_puns.html

Reply:
Aah cool it's like property shorthand in JS! It's a moot point because `WriteProcessedGameReport` has properties that don't belong in `ReadProcessedGameReport`, but is it possible to `spread` too? Like would there be an equivalent to:
```
{
    rid,
    winner,
    loser,
    ...report,
}
```
This is probably just JS giving me crazy ideas 😬

Reply:
I don't think that's a crazy idea at all - but Haskell syntax doesn't support it. There are a few similar things you can do though...

Suppose we have the types:
```hs
data Foo = {a :: Int, b :: Int, c :: Int}
data Bar = {a :: Int, b :: Int, d :: Int}
```
You can decompose a value into its fields with a pattern match, so we could write something like:
```hs
-- Naive, no pattern match
mkBar :: Foo -> Bar
mkBar foo = Bar {a = foo.a, b = foo.b, d = 7}

-- Pattern match version. Matches *any* Foo, but doesn't assign a name to `c`
mkBar' :: Foo -> Bar
mkBar' (Foo {a b}) = Bar {a, b, d = 7}
```

There's also record *update* syntax, which is handy when you want to change only a subset of the fields:
```hs
-- Take a Bar and change it's `a` field to match the Foo's `a` field. Bar's `b` and `d` stay untouched
updateBar :: Foo -> Bar -> Bar
updateBar (Foo {a}) bar = bar {a}
```


EDIT!!!!
In the process of writing this response, I learned about https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/record_wildcards.html

This has to be enabled as a language extension, but looks very similar to record spreading when used in conjunction with the above tricks. I'm going to look into this to see if it can save us some verbosity!

Reply:
Awesome find! Would be so cool if this pans out! (and the original pattern matching response is good learning for me too)